config/config_test.go

```diff
@@ -0,0 +1,333 @@
+	_, err = caFile.WriteString(testCA)
+	require.NoError(t, err)
+
+	cases := map[string]struct {
+		cfg        *Config
+		setupEnv   func() error
+		cleanupEnv func() error
+	}{
+		"TLS with CACertFile": {
+			cfg: &Config{
+				ConsulServers: ConsulServers{
+					Hosts:      "consul.dc1.address",
+					EnableTLS:  true,
+					CACertFile: caFile.Name(),
+				},
+			},
+		},
+		"TLS with CACertPEM": {
+			setupEnv: func() error {
+				return os.Setenv(consulCACertPemEnvVar, testCA)
+			},
+			cfg: &Config{
+				ConsulServers: ConsulServers{
+					Hosts:     "consul.dc1.address",
+					EnableTLS: true,
+				},
+			},
+			cleanupEnv: func() error {
+				return os.Unsetenv(consulCACertPemEnvVar)
+			},
+		},
+	}
+
+	for name, c := range cases {
+		t.Run(name, func(t *testing.T) {
+			if c.setupEnv != nil {
+				require.NoError(t, c.setupEnv())
+				t.Cleanup(func() {
+					require.NoError(t, c.cleanupEnv())
+				})
+			}
+			cfg, err := c.cfg.ConsulServerConnMgrConfig(awsutil.ECSTaskMeta{})
+			require.NoError(t, err)
+
+			require.NotNil(t, cfg.TLS.RootCAs)
+		})
+	}
+}
+
+func TestClientConfig(t *testing.T) {
+	caFile, err := os.CreateTemp("", "")
+	t.Cleanup(func() {
+		_ = os.RemoveAll(caFile.Name())
+	})
+	require.NoError(t, err)
+	_, err = caFile.WriteString(testCA)
+	require.NoError(t, err)
+
+	cases := map[string]struct {
+		cfg        *Config
+		expConfig  *api.Config
+		setupEnv   func() error
+		cleanupEnv func() error
+	}{
+		"basic flags without TLS": {
+			cfg: &Config{
+				ConsulServers: ConsulServers{
+					Hosts: "consul.dc1.address",
+				},
+			},
+			expConfig: &api.Config{
+				Scheme: "http",
+			},
+		},
+		"basic flags without TLS and custom service namespace and partition": {
+			cfg: &Config{
+				ConsulServers: ConsulServers{
+					Hosts: "consul.dc1.address",
+				},
+				Service: ServiceRegistration{
+					Name:      "test-service",
+					Namespace: "test-ns",
+					Partition: "test-par",
+				},
+			},
+			expConfig: &api.Config{
+				Scheme:    "http",
+				Namespace: "test-ns",
+				Partition: "test-par",
+			},
+		},
+		"basic flags without TLS and custom gateway namespace and partition": {
+			cfg: &Config{
+				ConsulServers: ConsulServers{
+					Hosts: "consul.dc1.address",
+				},
+				Gateway: &GatewayRegistration{
+					Name:      "test-service",
+					Namespace: "test-ns",
+					Partition: "test-par",
+				},
+			},
+			expConfig: &api.Config{
+				Scheme:    "http",
+				Namespace: "test-ns",
+				Partition: "test-par",
+			},
+		},
+		"TLS with CaCertFile": {
+			cfg: &Config{
+				ConsulServers: ConsulServers{
+					Hosts:      "consul.dc1.address",
+					EnableTLS:  true,
+					CACertFile: caFile.Name(),
+				},
+			},
+			expConfig: &api.Config{
+				Scheme: "https",
+				TLSConfig: api.TLSConfig{
+					Address: "consul.dc1.address",
+					CAFile:  caFile.Name(),
+				},
+			},
+		},
+		"TLS with CaCertPEM": {
+			setupEnv: func() error {
+				return os.Setenv(consulCACertPemEnvVar, testCA)
```

Comment:
minor: There is actually [`t.Setenv`](https://pkg.go.dev/testing#T.Setenv) we could use for this now